straxen/plugins/bayes_peaklet_classification.py

```diff
@@ -0,0 +1,141 @@
+
+    :param peaklets: peaklets
+    :param waveforms: peaklets waveforms in PE/ns
+    :param quantiles: quantiles in ns
+    :returns: the ln probability of a each peaklet belonging to S1 and S2 class
+    """
+    provides = 'peaklet_classification_bayes'
+    depends_on = ('peaklets',)
+    __version__ = '0.0.1'
+    dtype = (strax.peak_interval_dtype
+             + [('s1_ln_prob', np.float32, 'S1 ln probability')]
+             + [('s2_ln_prob', np.float32, 'S2 ln probability')]
+            )
+
+    # Descriptor configs
+    bayes_config_file = straxen.URLConfig(
+        default='resource://'
+        'conditional_probabilities_and_bins_v1_w_global_v6.npz?fmt=npy',
+        help='Bayes configuration file, conditional probabilities tables and Bayes discrete bins'
+    )
+    num_nodes = straxen.URLConfig(
+        default=50,
+        help='Number of attributes(features) per waveform and quantile'
+    )
+    classes = straxen.URLConfig(
+        default=2,
+        help='Number of label classes S1(1)/S2(2)'
+    )
+
+    def setup(self):
+
+        self.class_prior = np.ones(self.classes)/self.classes
+        self.bins = self.bayes_config_file['bins']
+        self.cpt = self.bayes_config_file['cprob']
+
+    def compute(self, peaklets):
+
+        waveforms, quantiles = compute_wf_and_quantiles(peaklets, self.num_nodes)
+
+        s1_ln_prob, s2_ln_prob = compute_inference(self.bins, self.num_nodes, self.cpt,
+                                                   self.classes, self.class_prior,
+                                                   waveforms, quantiles)
+
+        return dict(time=peaklets['time'],
+                    dt=peaklets['dt'],
+                    channel=-1,
+                    length=peaklets['length'],
+                    s1_ln_prob=s1_ln_prob,
+                    s2_ln_prob=s2_ln_prob
+                    )
+
+
+def compute_wf_and_quantiles(peaklets: np.ndarray, num_nodes: int):
+    """
+    Compute waveforms and quantiles for a given number of nodes(atributes)
+    :param peaklets:
+    :param num_nodes: number of nodes or atributes
+    :return: waveforms and quantiles
+    """
+    waveforms = np.zeros((len(peaklets), num_nodes))
+    quantiles = np.zeros((len(peaklets), num_nodes))
+
+    num_samples = peaklets['data'].shape[1]
+    step_size = int(num_samples/num_nodes)
+    steps = np.arange(0, num_samples+1, step_size)
+
+    data = peaklets['data'].copy()
+    dts = peaklets['dt'].copy()
+    data[data < 0.0] = 0.0
+    for i, p in enumerate(peaklets):
+        sample_number = np.arange(0, num_samples+1, 1)*p['dt']
+        frac_of_cumsum = np.append([0.0], np.cumsum(data[i, :]) / np.sum(data[i, :]))
+        cumsum_steps = np.interp(np.linspace(0., 1., num_nodes, endpoint=False), frac_of_cumsum, sample_number)
+        cumsum_steps = np.append(cumsum_steps, sample_number[-1])
+        quantiles[i, :] = cumsum_steps[1:] - cumsum_steps[:-1]
```

Comment:
This looks like a very intense operation with a for loop which is potentially quite slow, have you checked the performance of this plugin?

Reply:
Here are a few numbers on Dali
for an  average background run `CPU times: user 1min 15s, sys: 15.6 s, total: 1min 31s` 
for an average Rd run `CPU times: user 4min 59s, sys: 57.6 s, total: 5min 57s`
for an average Kr run `CPU times: user 8min 22s, sys: 1min 36s, total: 9min 59s`
which includes loading peaks

Reply:
comparing to run 12236, this seems much longer than any other plugin we had at that time I wrote this notebook:
https://github.com/XENONnT/workshop_zoom_2021_04_19/blob/master/1_xenonnt_data/1_Data_Structure_and_processing_tutorial.ipynb
![afbeelding](https://user-images.githubusercontent.com/22295914/155678742-bac19301-4b5a-412e-940a-2fc1d78a9863.png)

This being said, the times may differ a lot from run to run, so I don't know if the max ~350 s (total time) compares to the 600s wall-time you report here.

My suspicion is that this for-loop is probably a bottle neck. I  personally very much like [line-profiling ](https://pypi.org/project/line-profiler/) a lot in order to find bottlenecks and focus optimization on the parts of the code that are actually worth optimizing.

Let me know if you need some examples of how I would try and optimize for performance. A few general hints would concerning this block of code would be:
 - Figure out which part of the code takes the longest, if you spend a lot of time speeding something up that wasn't slow, you don't gain much (only because I've done this often myself, and it's a very tempting pitfall 😉 )
 - avoid for loops (unless you can do it in a numbafied function) or the numer of iterations is small
 - avoid `np.append`, every time you call that, you instruct numpy to allocate an entirely new block of memory which you then have to fill with the old stuff array

I can write you an example if you want, probably also useful for other people if @WenzDaniel or me are asking again and again if something is "fast" or "slow"

Reply:
Hi @JoranAngevaare 
Good point, at the moment this plugin is at the level of `event_pattern_fit` in terms of slowness. We already have an idea on how to remove `np.append` for the reason you mentioned, but we would need to retrain. 
If the priority is to have this ready so analyst can use it, I think we can go ahead and meanwhile we work on the optimization. Also I would like to get other people opinion @tunnell @tinapeters 

Btw, once I tried to run sneakviz  to check this but I end up lost looking at the profile because all straxen calls , do you have an example for dummies on how to do this for a single straxen plugin

Reply:
When running this plugin ~20% of the time is just loading peaks, I will try to debug more and see if can speed up 
```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    17                                           @profile
    18                                           def _compute_bayes(st):
    19                                           
    20         1      45984.0  45984.0      0.0      p = st.get_single_plugin('031711', 'peak_classification_bayes')
    21                                           
    22         1   21889439.0 21889439.0     21.0      peaks = st.get_array('031711', 'peaks')
    23                                               
    24         1   82340242.0 82340242.0     79.0      result =p.compute(peaks)

```